pkg/api/defaults-custom-cloud-profile.go

```diff
@@ -93,100 +110,114 @@ func (cs *ContainerService) SetCustomCloudProfileEnvironment() error {
+			}
 		}
 	}
+
 	return nil
 }
 
-// AzureStackCloudSpecParams is the parameters when we set the azure stack cloud spec defaults for ContainerService.
-type AzureStackCloudSpecParams struct {
+// AzureCustomCloudSpecParams is the parameters when we set the azure custom cloud spec defaults for ContainerService.
+type AzureCustomCloudSpecParams struct {
 	IsUpgrade bool
 	IsScale   bool
 }
 
-// SetAzureStackCloudSpec sets the cloud spec for Azure Stack .
-func (p *Properties) SetAzureStackCloudSpec(params AzureStackCloudSpecParams) error {
-	if p.IsAzureStackCloud() {
-		var azureStackCloudSpec AzureEnvironmentSpecConfig
+// SetAzureCustomCloudSpec sets the cloud spec for Custom Cloud .
+func (p *Properties) SetCustomCloudSpec(params AzureCustomCloudSpecParams) error {
+	if p.IsCustomCloudProfile() {
+		var azureCustomCloudSpec AzureEnvironmentSpecConfig
 		switch p.CustomCloudProfile.DependenciesLocation {
-		case AzureStackDependenciesLocationPublic:
-			azureStackCloudSpec = AzureCloudSpecEnvMap[AzurePublicCloud]
-		case AzureStackDependenciesLocationChina:
-			azureStackCloudSpec = AzureCloudSpecEnvMap[AzureChinaCloud]
-		case AzureStackDependenciesLocationGerman:
-			azureStackCloudSpec = AzureCloudSpecEnvMap[AzureGermanCloud]
-		case AzureStackDependenciesLocationUSGovernment:
-			azureStackCloudSpec = AzureCloudSpecEnvMap[AzureUSGovernmentCloud]
+		case AzureCustomCloudDependenciesLocationPublic:
+			azureCustomCloudSpec = AzureCloudSpecEnvMap[AzurePublicCloud]
+		case AzureCustomCloudDependenciesLocationChina:
+			azureCustomCloudSpec = AzureCloudSpecEnvMap[AzureChinaCloud]
+		case AzureCustomCloudDependenciesLocationGerman:
+			azureCustomCloudSpec = AzureCloudSpecEnvMap[AzureGermanCloud]
+		case AzureCustomCloudDependenciesLocationUSGovernment:
+			azureCustomCloudSpec = AzureCloudSpecEnvMap[AzureUSGovernmentCloud]
 		default:
-			azureStackCloudSpec = AzureCloudSpecEnvMap[AzurePublicCloud]
+			azureCustomCloudSpec = AzureCloudSpecEnvMap[AzurePublicCloud]
 		}
 		if p.CustomCloudProfile.Environment == nil || p.CustomCloudProfile.Environment.ResourceManagerVMDNSSuffix == "" {
 			return errors.New("Failed to set Cloud Spec for Azure Stack due to invalid environment")
 		}
 
-		azureStackCloudSpec.EndpointConfig.ResourceManagerVMDNSSuffix = p.CustomCloudProfile.Environment.ResourceManagerVMDNSSuffix
-		azureStackCloudSpec.CloudName = AzureStackCloud
+		azureCustomCloudSpec.EndpointConfig.ResourceManagerVMDNSSuffix = p.CustomCloudProfile.Environment.ResourceManagerVMDNSSuffix
+
+		if (p.CustomCloudProfile.Environment.Name == "" || p.IsAzureStackCloud()) {
+			azureCustomCloudSpec.CloudName = AzureStackCloud
+		} else {
+			azureCustomCloudSpec.CloudName = p.CustomCloudProfile.Environment.Name
+		}
 
 		//Sets default values for telemetry PID where none is set
 		if p.CustomCloudProfile.AzureEnvironmentSpecConfig == nil {
```

Comment:
those logic are Azure Stack only

Reply:
Would it break if I set default values to empty string for non-AzureStack custom clouds?

If so, I actually do not think there is any harm for defaulting AzureTelemetryPID to AZureStackTelemetyPIDs.

Reply:
I am not sure that if the empty id is an issue for non-auzre stack environment, 
the behavior is controlled by ARM, 
Azure Stack ARM has special logic to target specific guids to send telemetry data.